src/System.Management.Automation/engine/hostifaces/PowerShell.cs

```diff
@@ -3083,10 +3083,8 @@ public IAsyncResult BeginInvoke<T>(PSDataCollection<T> input, PSInvocationSettin
         /// <exception cref="ObjectDisposedException">
         /// Object is disposed.
         /// </exception>
-        public async Task<PSDataCollection<PSObject>> InvokeAsync()
-        {
-            return await Task<PSDataCollection<PSObject>>.Factory.FromAsync(BeginInvoke(), _endInvokeMethod).ConfigureAwait(false);
-        }
+        public Task<PSDataCollection<PSObject>> InvokeAsync()
+            => Task<PSDataCollection<PSObject>>.Factory.FromAsync(BeginInvoke(), _endInvokeMethod);
```

Comment:
The thing I'm curious about is that the documentation [here](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.taskfactory.fromasync?view=netframework-4.7.2#System_Threading_Tasks_TaskFactory_FromAsync_System_IAsyncResult_System_Action_System_IAsyncResult__) says that the `FromAsync` overload taking an `IAsyncResult` rather than a `Func<AsyncCallback, object, IAsyncResult>` is slower.

I tried looking into this, and the doc written for wrapping the old API with the new one [here](https://docs.microsoft.com/en-us/dotnet/standard/parallel-programming/tpl-and-traditional-async-programming), but I'm not quite sure what needs to be done to turn the `BeginInvoke()` method we have into the appropriate callback handler

Reply:
[Here](https://stackoverflow.com/questions/31764225/why-do-taskfactory-fromasync-overloads-require-a-state-object-to-be-supplied)'s an SO post that touches on this question

Reply:
[Here](https://stackoverflow.com/questions/6313114/performance-cost-of-using-the-iasyncresult-overload-of-task-fromasync)'s another one that has some more detail

Reply:
Ah, after reading a bit more, it looks like the callback functionality would need to be supported by PowerShell itself and passed into a `BeginInvoke()` method we provide so that it can signal that the task is done.

I was proceeding under the impression that it might be possible to fudge that with a wrapping lambda based on the documentation's recommendation of "use this overload", but I'm guessing that's wrong.

Reply:
Yeah, our Begin/End implementation doesn't strictly conform to the APM contract.

Reply:
Will improve?

Reply:
I don't see a reason to update the existing APM APIs.

Reply:
There was nor addressed comment https://github.com/PowerShell/PowerShell/pull/8056#issuecomment-468545375

Reply:
I think we can defer opening that issue until we get some real feedback on those async APIs that request performance enhancement.
Bruce brought that up mainly because when it was discussed in the committee meeting, there was an ask about if we should have separate implementation for the task-based APIs.